Objects/listobject.c

```diff
@@ -928,10 +933,14 @@ list_extend(PyListObject *self, PyObject *iterable)
          * eventually run out of memory during the loop.
          */
     }
+    else if (self->ob_item == NULL) {
+        /* We have to rely on n being correct in this case. */
```

Comment:
This comment seems misleading. It's OK for the `n` returned by `PyObject_LengthHint` to be incorrect. We're pre-allocating enough memory for `n` extra elements, but it's still OK for us to find fewer or more than `n` elements when we iterate over the iterable below.

We only need to skip the call to `list_preallocate_exact` when `n == 0` because `list_preallocate_exact` enforces the invariant that `n > 0` (and indeed, it doesn't make sense to call it to allocate zero bytes of memory).

Reply:
Removed.